### ISSUE TITLE
Support Zig as backend C compiler 

### ISSUE THREAD

**Issue by jimkring** [Nuitka/Nuitka#2741]
zig is a nice language and toolchain.

The zig compiler supports C and C++ compilation (by building clang into zig) — the `zig cc` command is the C compiler  and `zig c++` is the C++ compiler.

it’s a single download with no dependencies.

It is currently possible to “fool” Nuitka into using zig by creating a shell script named “zig-clang-cc.sh” that passes all command line arguments to `zig cc` and setting `CC=“/path/to/zig-clang-cc.sh”`. Note, it is also necessary to pass the `--clang` option to nuitka.

I set up a [simple project](https://github.com/jimkring/nuitka-zig-cc) with [GitHub action build](https://github.com/jimkring/nuitka-zig-cc/actions/workflows/ci.yml) to test this approach.

initial testing quickly found [a bug in zig’s handling of some linker flags](https://github.com/ziglang/zig/pull/19176).

it would be nice to add support to nuitka for officially detecting when zig is configured as the compiler — when CC is set to `zig cc`.

Note that full support will probably not work right away, or soon, but it would be good to try to keep improving integration and share bug reports with the zig project, because there are many long-term benefits for using zig.

**Comment by kayhayen** [comment]
I agree, adding that will make it much easier to track the progress on zig side. Notice however that due to https://github.com/ziglang/zig/issues/11415 this may be a very long time away actually.

For older than 3.11, I think we might make it work with using C++03 workaround. Unfortunately, no way to compile for 3.11 without the C11 support there. But ever since MSVC can do C11 now, maybe it's gaining more traction, as e.g. seen in Python.

**Comment by jimkring** [comment]
I have created this tracking issue for communicating our Zig issue priorities to their team: https://github.com/Nuitka/Nuitka/issues/2742

**Comment by polarathene** [comment]
> It is currently possible to “fool” Nuitka into using zig by creating a shell script named `zig-clang-cc.sh` that passes all command line arguments to `zig cc` and setting `CC=/path/to/zig-clang-cc.sh`.

This is no longer the case.

> For older than 3.11, I think we might make it work with using C++03 workaround. Unfortunately, no way to compile for 3.11 without the C11 support there.

With workarounds to get Zig working with current Nuitka release, this statement is a bit unclear. I was able to build with Zig successfully, but I've only tried with a basic hello world project whilst filing detailed reports 😅

~~I'm assuming this information is now potentially outdated, despite the open issue at Zig, perhaps that has been resolved?~~ (**EDIT:** [Nope](https://github.com/ziglang/translate-c/issues/149#issuecomment-3394041743))

I guess it's to do with compiling libpython from source perhaps? I'm not sure why you're blocked by `zig translate-c`, do you need Zig source, or is `zig cc` sufficient?

**Comment by kayhayen** [comment]
In Nuitka there is this code to support zig:

```        
if isZigName(env.the_cc_name):
    values.append("c++" if env.c11_mode else "cc")
```

I was never going to use the workaround of couse.

**Comment by kayhayen** [comment]
@polarathene Which version of zig and Python were you using?

**Comment by kayhayen** [comment]
> I agree, adding that will make it much easier to track the progress on zig side. Notice however that due to [ziglang/translate-c#149](https://github.com/ziglang/translate-c/issues/149) this may be a very long time away actually.

That point of mine was misguided, that is a tool to translate C to zig, which doesn't have to do that at all.

**Comment by polarathene** [comment]
Apologies it's taking so long to write up the bug report 😓 I should have it finished with plenty of details later today.

---

> In Nuitka there is this code to support zig:
> 
> ```
> if isZigName(env.the_cc_name):
>     values.append("c++" if env.c11_mode else "cc")
> ```

Yeah I'm aware of logic like this in the Nuitka source :)

IIRC Nuitka was calling `zig c++` when I inspected crash reports while troubleshooting Zig support.

---

> Which version of zig and Python were you using?

Zig 14.1 and 15.1 with Python 3.11 (Fedora package and `python-build-standalone` from UV).

Additional changes were required to get 3.12 and 3.13 to build successfully. There was 3 lines of C from Nuitka that Zig complained about but were all fixable with a static type cast.

I was also able to cross-compile from `x86_64` to `aarch64` without needing a separate toolchain for cross compilation setup. This also includes cross-compiling with glibc at whatever version targeted for avoiding a legacy build environment.

Separately I also tried the `wasm32-wasi` target from Zig and this had header file issues `bits/wordsize.h` and `pyconfig-32.h` not being available. Presumably if those were supplied correctly then this target would also work.

**Comment by kayhayen** [comment]
I need to implement `--compiler` as a replacement for the current mess, as it's coming pretty untenable to deal with it as a bunch of booleans and also boolean command line options. But when set as CC to full path of zig, current factory branch does it fine now for Linux and Windows. I will look at macOS later.

**Comment by kayhayen** [comment]
Thanks for your report. This is fixed on the factory branch, which is a development version that is under rapid development. You can try it out by going here: https://nuitka.net/doc/factory.html 
	
Feedback on whether this is working is very welcome. Please do not share plans to do it; only confirm or deny that it is working.

**Comment by kayhayen** [comment]
C11 works just fine now, no C++ workaround needed anymore. Even C23 embedding seems to work fine. I have to add a TODO for the disabling of the caching, which apparently is not possible with zig, but that I know professional users prefer to have. But there is probably only so much to do. Probably error out and point to `--clean-cache=ccache` as a workround.

**Comment by kayhayen** [comment]
The changes I made are part of the current pre-release of Nuitka. It will indeed be super interesting to see if zig compiles faster, but for now it probably won't qualify as a time saver for myself given that I have encountered these unwarranted panics in debug mode.

**Comment by jimkring** [comment]
@kayhayen are there docs that describe the steps to use zig as the compiler for Nuitka?  If no docs exists and it’s trivial for you to describe those steps here, would you be willing to details those out here in a reply?  Thanks either way.

**Comment by jimkring** [comment]
> The changes I made are part of the current pre-release of Nuitka. It will indeed be super interesting to see if zig compiles faster, but for now it probably won't qualify as a time saver for myself given that I have encountered these unwarranted panics in debug mode.

Regarding the "faster compiles" thought, (maybe you already know this)...

The latest Zig ([as of around June 2025](https://ziglang.org/devlog/2025/#2025-06-08)) uses its new, self-hosted x86 backend by default on x86_64 Linux (and soon, but I don't know when on Windows), which is purported to be much faster than the LLVM compiler (which is still used for release builds).

**Comment by kayhayen** [comment]
So what I did was to download from https://ziglang.org/download/ the latest stable release, currently 15.2, we don't dare development releases obviously and to set the CC environment. Since I use bash, I do this typically like so:

> CC=~/test/zig-x86_64-windows-0.15.2/zig.exe python3.12 bin/nuitka-run --onefile tests/basics/AssertsTest.py

Later, I envision that I am going to finally add support for `--compiler=zig` which would pick it up from the system PATH, but due to scalability work having priority right now, I am not going there yet. That new option will be quite diffiicult to transition.

This is something that is supposed to work in Nuitka since forever, I mean the CC environment. It then guesses the compiler type by basename of the file.

The `--version` output suffers from this:

> Version C compiler: ~\test\zig-x86_64-windows-0.15.2\zig.exe (zig.exe not found).

But I am going to ignore that for now. Probably means that I ought to add the dirname to PATH, but I am supprised where it is getting the idea to use the basename for version detection anyway.

Already I am scared of the regressions this may have caused, although I am thinking that it also got warnings of ClangCL sorted out, and to be honest, what I did was mostly to remove repetitive code from our 4 scons files that deals with CC and sometimes CXX, and not as much zig support itself.

Regaring the other comment, the self hosted backend might explain why it crashes only on Linux @jimkring and I actually thought that "zig cc" was already fully self sustained, but I didn't follow the lower technical levels, but I thought that was why cross compilation was relatively easy for them, but that's not that important. If it's still new that might explain why it has bugs.

I will add it to my CI once I have the `--compiler` setting implemented, and will then report issues I find. But this was a crasher with a first hello world style program. Not too encouraging. Later the crash changed places, but remained similar kind of assertion of a value to be None that is likely not.

And at some point, I might even make it download zig automatically much like I do with winlibs. But we still need `gdb/lldb` from there on Windows for tooling, and maybe other things.

Anyway, I expect to make a test on macOS still and then consider this issue by itself settled for now. At some point I will add --compiler secretly, and make it conflict with `--clang`, `--msvc`, `--mingw64` and make it work for some things, but not everything, I would e.g. love to see things like `--compiler=system-gcc` and `--compiler=gcc>14` and things like `--compiler=msvc>=14,2,gcc>=15,conda-cc` to be implemented where the later attempts to use  a compiler from a list, does version checks on them, and error exits if no matching one is found. But that's going to cost me days which I currently don't want to spend really.

**Comment by kayhayen** [comment]
So on macOS, it hates the value "arm64" as part of `--target` which may or may not be necessary anyway, so I removed that for now, it seems they name the architecture different from what Apple does, but we can use the native architecture there just fine I believe.

And then I had an error with how LTO is enabled, which gives errors. Apparently on macOS zig doesn't support `-flto` due to the linker choice. This might be something that can be enhanced if we figure out how to make it use ldd, but apparently on macOS the system linker is to be used.

Then I had

> error: unsupported linker arg: -sectcreate 

I have another pending action there, namely that is to use C23 embedding in favor of sections for the constant blobs used on macOS with new enough compiler. I might be wrong, but for macOS the Clang 17 didn't do it yet, only Clang 19 would, and macOS 26 would have that only. But for zig, it would be best to use it there too. That will allow to get rid of having to support platform specific code there a bunch. However, since I did do it for ClangCL on Windows already, I could make it be used for all OSes with zig, and then it works.

And I got a hello world with "--debug", but *without* `--debug`  I am getting this:

> Nuitka: Launching 'AssertsTest.bin'.
SystemError: /Users/nad/build_macos_installer/installer/variant/binaries/build_source/Objects/codeobject.c:455: bad argument to internal function
Abort trap: 6

That's an assertion from the Python DLL, which may actually be true potentially, in case of uninitialized data being used, but I am somehow sceptical of it right now, but I do think it's worth checking.

**Comment by kayhayen** [comment]
I am getting the idea that maybe the crashes I saw on Linuxx and this one are linked, the constants blob code reading code objects, may not be portable when it reads data into integers perhaps, that kind of thing.

**Comment by kayhayen** [comment]
I added a bunch of assertions, which of course won't be used though, and aborts for some values, to no avail. I think the next line of action there is to go and use Python in a self-compiled mode. The LTO level made it kind of impossible to tell what part of the assertions of libpython are triggering, it's just a big single one on many things. I tried to check all of them, but apparently that doesn't fire.

**Comment by kayhayen** [comment]
Self compiling on macOS is also one thing I haven't done before, and I got other more important stuff to do for now, so I am putting this to delayed, I cannot let this take precendence over scalability work this month.

**Comment by kayhayen** [comment]
So, I took a timeout and actually did the self compiled Python for macOS in what I consider the best way. However, zig dislikes the libpython.a and talks about this for all object files in there:

```
    note: while parsing /Users/hayen/test/macos-self-compile/Python-3.11.9/libpython3.11.a(call.o)
error: unknown cpu architecture: 0  
```

Not sure how the arch mismatch can be resolved, they seems to use aarch64 and apple does arm64 and I don't know how to reconcile it. Only thing is I could make a version that is either "fat", i.e. machine code, and maybe that works better than bytecode for this purpose. But it wasn't going to be as easy I guess.

**Comment by kayhayen** [comment]
When adding proper version output for zig (which also revealed that the version for Clang was not detected from the actual compiler, but via PATH always only) I noticed I had been using 0.16 dev rather than a stabel zig release and turns out that macOS is just running fine the basic tests now with zig. For the self compiled macOS script, I won't have to put it to use now.

**Comment by kayhayen** [comment]
In debug mode, they find this:

> thread 29920873 panic: variable length array bound evaluates to non-positive value 0
???:?:?: 0x1000cd12f in _Nuitka_CallFunctionPosArgs (???)

**Comment by kayhayen** [comment]
That's when a function with no argument gets a positional only call with arguments, which other C compilers take more lightly.

**Comment by jimkring** [comment]
@kayhayen FYI the “recommended” way to install zig for integration with Python is via `pip install ziglang`. I’m not sure how that plays out here, but it’s worth considering, since it could simplify steps for setup and use of zig by Nuitka.

**Comment by kayhayen** [comment]
Oh, that will actually be great, since that would mean that I don't have to implement any download capability (or can do so via pip install), great to know. I guess that also adds it to the PATH of a virtualenv.

**Comment by kayhayen** [comment]
I managed to workaround that zig limitation (it's a language extension to allow zero length array) and now macOS just works fine.

**Comment by kayhayen** [comment]
Linux passes basic tests, but panics when ran with `--debug`, we are going to ignore that for now. That's probably an area where `zig cc` is not yet as bug free.

**Comment by kayhayen** [comment]
The use of CC=zig ought to be sufficient with pip install then, btw, no need for full paths if it's in PATH. I didn't install it on a system path.

I have started some work where Nuitka downloads things like orderedset too into a private area, which it and only it sees at runtime, but that experiment got stuck. But in principle that would be a nice solution, if Nuitka had that cached somewhere and shared across versions. Then adding --compiler=zig support would in fact be relatively trivial as in "zig not in path" -> "pip install to private cache folder" and use it from there.

**Comment by kayhayen** [comment]
All the necessary changes are on the factory branch now.

**Comment by kayhayen** [comment]
Part of the current develop branch again, I am removing the delayed tag, as it seems I have done it anyway.

**Comment by jimkring** [comment]
@kayhayen it worked perfectly for me! Here are the steps I took (on Windows), for everyone's benefit.

- used `uv init` to create a default python+uv project in my new/empty `C:\Projects\nuitka-testing` project directory.
- changed the python version to 3.12 in both the `.python-version` and `pyproject.toml` files
- used `uv venv` to create the python 3.12 virtual environment.
- added `nuitka` factory and `ziglang` to my `pyproject.toml` dependencies as shown below:
```toml
[project]
name = "nuitka-testing"
version = "0.1.0"
description = "Add your description here"
readme = "README.md"
requires-python = ">=3.12"
dependencies = [
    "nuitka@https://github.com/Nuitka/Nuitka/archive/factory.zip",
    "ziglang>=0.15.1",
]
```
- used `uv sync` to install all the dependencies
- activated the `.venv` with `.\.venv\scripts\activate`
- set `CC` env variable to "zig" (via `$env:CC = "zig"` and `$CC = "zig"` for safe measure)
- used `nuitka --onefile hello.py` to build the `hello.py` file (that `uv init` creates) into an executable
- saw that nuitka knows it's using zig as its compiler:
```
(nuitka-testing) PS C:\Projects\nuitka-testing> nuitka --onefile hello.py
Nuitka-Options: Used command line options:
Nuitka-Options:   --onefile hello.py
Nuitka: Starting Python compilation with:
Nuitka:   Version '2.9rc4' on Python 3.12 (flavor 'Unknown') commercial grade 'not installed'.
Nuitka: Completed Python level compilation and optimization.
Nuitka: Generating source code for C backend compiler.
Nuitka: Running data composer tool for optimal constant value handling.
Nuitka: Running C compilation via Scons.
Nuitka-Scons: Backend C compiler: zig (zig 0.14.0).
Nuitka-Scons: Backend C linking with 6 files (no progress information available for this stage).
Nuitka-Onefile: Creating single file from dist folder, this may take a while.
Nuitka-Onefile: Running bootstrap binary compilation via Scons.
Nuitka-Scons: Onefile C compiler: zig (zig 0.14.0).
Nuitka-Scons: Onefile C linking.
Nuitka-Onefile: Using compression for onefile payload.
Nuitka-Onefile: Onefile payload compression ratio (25.88%) size 17342666 to 4487950.
Nuitka-Onefile: Keeping onefile build directory 'hello.onefile-build'.
Nuitka: Keeping dist folder 'hello.dist' for inspection, no need to use it.
Nuitka: Keeping build directory 'hello.build'.
Nuitka: Successfully created 'C:\Projects\nuitka-testing\hello.exe'.
```
- ran the resulting exe to ensure that it works:
```
(nuitka-testing) PS C:\Projects\nuitka-testing> .\hello.exe
Hello from nuitka-testing!
```

**Comment by jimkring** [comment]
@kayhayen and @KRRT7 -- Would it add value to include such how-to instructions in Nuitka's official docs/site/wiki so that interested parties can benefit from it?  My thinking is that this is a really valuable feature for Nuitka users and the python community in general.

**Comment by jimkring** [comment]
> [@kayhayen](https://github.com/kayhayen) and [@KRRT7](https://github.com/KRRT7) -- Would it add value to include such how-to instructions in Nuitka's official docs/site/wiki so that interested parties can benefit from it? My thinking is that this is a really valuable feature for Nuitka users and the python community in general.

@KRRT7 what is the best way to get such information into the Nuitka docs?

**Comment by KRRT7** [comment]
@dauryjey can you add it on to your todo list please?

**Comment by kayhayen** [comment]
I personally much rather than having docs would implement the `--compiler` option. If I have the option to write docs, and to remove the need for it, I know what to do with it is going to be to make this automatic. So that `--compiler=zig` just works out of the box using a zig in PATH (if it has sufficient version) or download it and use it automatically. Had I done docs for how to install MinGW64 and I vaguely recall doing that, I was tormented with users incapable of selecting an architecture and so, until I made it automatic, then only the semi pros that didn't read and thought their self installed MinGW ought to match when it didn't, was that.

At this time I am a bit -1 to add anything to the docs before I haven't see that it is any good. For experimental support of zig, what we could and should maybe do, is to write a blog post about it and publish one for once. I might use some AI help with the wording and do that indeed. It's about time, and this is in fact newsworthy stuff.

For now, my default for compiler would be on Windows, `--compiler=msvc,mingw64` until I am convinced that zig is not a degradation, and as stated, I lack the time completely, as I actually didn't even mean to implement this. I might be going to `--compiler` to clean up the compiler selection process a bunch though. Having to do --msvc --clang for clangcl and --mingw --clang for clang from winlibs, it's kind of crazy, and nothing that should have docs too.

If @dauryjey were to spend time on this, I guess he could try his hand at --compiler, but it's complicated as in (enormous amound of details to be touched) so I think that would be a waste of time.

But what if you wrote that blog post daury? I would say it ought to explain how Nuitka uses a C compiler for generating the actual machine code, and then point out the various supported ones in the existing docs, and that with the next Nuitka release (4.0) we are going to add zig support at least experimentally, and that we still have to run our test suites with all zig.

Reminds me, did anybody do a self compiled Python on Windows with zig yet? Is that even possible?

**Comment by fleimgruber** [comment]
> Reminds me, did anybody do a self compiled Python on Windows with zig yet? Is that even possible?

The Zig tagline is "Zig is a general-purpose programming language and toolchain for maintaining robust, optimal and reusable software." and they take the word "toolchain" seriously so they ported CPython to the Zig build system: https://github.com/allyourcodebase/cpython.

I have not tried it on Windows yet, but am going to do that right now.

**Comment by jimkring** [comment]
Interoperability with Zig, even if early/experimental, will make a huge different for Nuitka and the Python community, IMO. I've been following it closely. They are doing some revolutionary work in C toolchain work.

**Comment by kayhayen** [comment]
You got me convinced. I am a huge zig fan anyway. Deeply impressed by toolchain and zig as a language. I will try and sneak in some time to work on this once I completed my next mini catchup. I have a grand vision of `--compiler` but honestly it doesn't have to be feature complete right away for it to become useful, and a 2.9 might appear sooner than I was thinking, if I can complete my next scalability step for real, although I mean to call it 4.0 for sanity reasons, skipping the Python version number space.

**Comment by kayhayen** [comment]
I have been working on the zig support some more, and decided to let go of my `--compiler` switch idea, as it requires way too much refactoring, and on the other hand, on windows, where MSVC is something that was previously backed by easy MinGW64 availability, now with 3.13+ that's not really easy anymore at all, so I am adding zig as the fallback in case MSVC is not available, and there is going to be a new command line `--zig`.

These changes should hit develop soon enough as well and will resolve this issue entirely. I refacted the code I have had for private pip space somewhat.

One complication is that zig is not supported directly by Scons, but there is code to just manually setup the environment, that will work it seems. I think previously I have used only replacements from `env["CC"]`, but this ought to be better.

**Comment by kayhayen** [comment]
This is now on the factory branch, but I didn't actually test if it does the expected fallback, since I don't really have a non-MSVC system available, I will only be able to test it through hacks, or maybe via using the wrong MSVC prompt or something like that. I will appreciate feedback. I didn't do widescale testing of zig, consider the use experimental.

**Comment by AxelVoitier** [comment]
Hello,

I happen to need to do a windows compilation for once, and since I use python 3.13 and don't have MSVC installed on the machine I have at my disposal, I though I would give zig a try.

 First issue I encounter is that nuitka does not seem to find the compiler at all. I have set up a clean venv, with Nuitka@factory and ziglang installed, and I get these:
```
(.venv) PS C:\Users\xxxx\Documents\nuitka-zig> python -m pip list
Package     Version
----------- -------
Nuitka      4.0rc2
ordered-set 4.1.0
pip         25.3
ziglang     0.15.1
zstandard   0.25.0
(.venv) PS C:\Users\xxxx\Documents\nuitka-zig> python-zig version
0.15.1
(.venv) PS C:\Users\xxxx\Documents\nuitka-zig> python -m nuitka .\hello.py
Nuitka-Options: Used command line options:
Nuitka-Options:   --onefile --zig .\hello.py
Nuitka: Starting Python compilation with:
Nuitka:   Version '4.0rc2' on Python 3.13 (flavor 'CPython Official') commercial grade 'not installed'.
Nuitka: Completed Python level compilation and optimization.
Nuitka: Generating source code for C backend compiler.
Nuitka: Running data composer tool for optimal constant value handling.
Nuitka: Running C compilation via Scons.
Nuitka depends on 'ziglang' to compile code, it is recommended.

Is it OK to download and put it in local user cache.

Fully automatic, cached. Proceed and download? [Yes]/No : no
FATAL: Nuitka needs to use 'ziglang'.
(.venv) PS C:\Users\xxxx\Documents\nuitka-zig> $env:CC = "python-zig"
(.venv) PS C:\Users\xxxx\Documents\nuitka-zig> python -m nuitka .\hello.py
Nuitka-Options: Used command line options:
Nuitka-Options:   --onefile --zig .\hello.py
Nuitka: Starting Python compilation with:
Nuitka:   Version '4.0rc2' on Python 3.13 (flavor 'CPython Official') commercial grade 'not installed'.
Nuitka: Completed Python level compilation and optimization.
Nuitka: Generating source code for C backend compiler.
Nuitka: Running data composer tool for optimal constant value handling.
Nuitka: Running C compilation via Scons.
Nuitka depends on 'ziglang' to compile code, it is recommended.

Is it OK to download and put it in local user cache.

Fully automatic, cached. Proceed and download? [Yes]/No : no
FATAL: Nuitka needs to use 'ziglang'.
(.venv) PS C:\Users\xxxx\Documents\nuitka-zig> $CC = "python-zig"
(.venv) PS C:\Users\xxxx\Documents\nuitka-zig> python -m nuitka .\hello.py
Nuitka-Options: Used command line options:
Nuitka-Options:   --onefile --zig .\hello.py
Nuitka: Starting Python compilation with:
Nuitka:   Version '4.0rc2' on Python 3.13 (flavor 'CPython Official') commercial grade 'not installed'.
Nuitka: Completed Python level compilation and optimization.
Nuitka: Generating source code for C backend compiler.
Nuitka: Running data composer tool for optimal constant value handling.
Nuitka: Running C compilation via Scons.
Nuitka depends on 'ziglang' to compile code, it is recommended.

Is it OK to download and put it in local user cache.

Fully automatic, cached. Proceed and download? [Yes]/No : no
FATAL: Nuitka needs to use 'ziglang'.
```

Finally let it download it itself leads to another kind of non-detection:
```
(.venv) PS C:\Users\xxxx\Documents\nuitka-zig> python -m nuitka .\hello.py
Nuitka-Options: Used command line options:
Nuitka-Options:   --onefile --zig .\hello.py
Nuitka: Starting Python compilation with:
Nuitka:   Version '4.0rc2' on Python 3.13 (flavor 'CPython Official') commercial grade 'not installed'.
Nuitka: Completed Python level compilation and optimization.
Nuitka: Generating source code for C backend compiler.
Nuitka: Running data composer tool for optimal constant value handling.
Nuitka: Running C compilation via Scons.
Nuitka depends on 'ziglang' to compile code, it is recommended.

Is it OK to download and put it in local user cache.

Fully automatic, cached. Proceed and download? [Yes]/No : yes
Requirement already satisfied: ziglang in c:\users\xxxx\documents\nuitka-zig\.venv\lib\site-packages (0.15.1)
FATAL: Error, failure to provide argument for 'zig_exe_path', please report bug.
```

Note that oddly enough, I seem to have gotten the zig compiler detected in another venv. But I was having another issue there (somehow related to `--onefile` maybe). And while I was trying to isolate a minimal reproducer in a clean venv, that's how I got to this non-detection issue.

**Comment by AxelVoitier** [comment]
So, got it to next step by uninstalling ziglang, and let nuitka install it itself. It does proceed to compilation step.

But now I encounter the following issue:
```
(.venv) PS C:\Users\xxxx\Documents\nuitka-zig> python -m pip uninstall ziglang
Found existing installation: ziglang 0.15.1
Uninstalling ziglang-0.15.1:
  Would remove:
    c:\users\xxxx\documents\nuitka-zig\.venv\lib\site-packages\ziglang-0.15.1.dist-info\*
    c:\users\xxxx\documents\nuitka-zig\.venv\lib\site-packages\ziglang\*
    c:\users\xxxx\documents\nuitka-zig\.venv\scripts\python-zig.exe
Proceed (Y/n)? y
  Successfully uninstalled ziglang-0.15.1
(.venv) PS C:\Users\xxxx\Documents\nuitka-zig> python -m nuitka .\hello.py
Nuitka-Options: Used command line options:
Nuitka-Options:   --onefile --zig .\hello.py
Nuitka: Starting Python compilation with:
Nuitka:   Version '4.0rc2' on Python 3.13 (flavor 'CPython Official') commercial grade 'not installed'.
Nuitka: Completed Python level compilation and optimization.
Nuitka: Generating source code for C backend compiler.
Nuitka: Running data composer tool for optimal constant value handling.
Nuitka: Running C compilation via Scons.
Nuitka depends on 'ziglang' to compile code, it is recommended.

Is it OK to download and put it in local user cache.

Fully automatic, cached. Proceed and download? [Yes]/No :
Collecting ziglang
  Using cached ziglang-0.15.1-py3-none-win_amd64.whl.metadata (3.2 kB)
Using cached ziglang-0.15.1-py3-none-win_amd64.whl (94.0 MB)
Installing collected packages: ziglang
  WARNING: The script python-zig.exe is installed in 'C:\Users\xxxx\Documents\nuitka-zig\.venv\Lib\site-packages\nuitka\build\inline_copy\downloads\pip\Users\xxxx\Documents\nuitka-zig\.venv\Scripts' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
Successfully installed ziglang-0.15.1
Nuitka-Scons: Backend C compiler: ~\Documents\nuitka-zig\.venv\Lib\site-packages\nuitka\build\inline_copy\downloads\pip\Users\xxxx\Documents\nuitka-zig\.venv\Lib\site-packages\ziglang\zig.exe (zig.exe not found).
Nuitka-Scons: Backend C linking with 7 files (no progress information available for this stage).
Nuitka-Inclusion:WARNING: Error, cannot detect used DLLs for DLL 'hello.dist\hello.dll' in package 'None' due to: expected str, bytes or os.PathLike object, not bool
Detecting used DLLs:  33.3%|████████▎                | 1/3 DLLs - hello.dllTraceback (most recent call last):
  File "C:\Users\xxxx\Documents\nuitka-zig\.venv\Lib\site-packages\nuitka\__main__.py", line 214, in <module>
    main()
    ~~~~^^
  File "C:\Users\xxxx\Documents\nuitka-zig\.venv\Lib\site-packages\nuitka\__main__.py", line 196, in main
    nuitka_main()
    ~~~~~~~~~~~^^
  File "C:\Users\xxxx\Documents\nuitka-zig\.venv\Lib\site-packages\nuitka\MainControl.py", line 1296, in main
    _main()
    ~~~~~^^
  File "C:\Users\xxxx\Documents\nuitka-zig\.venv\Lib\site-packages\nuitka\MainControl.py", line 1158, in _main
    detectUsedDLLs(
    ~~~~~~~~~~~~~~^
        standalone_entry_points=getStandaloneEntryPoints(),
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    ...<2 lines>...
        ),
        ^^
    )
    ^
  File "C:\Users\xxxx\Documents\nuitka-zig\.venv\Lib\site-packages\nuitka\freezer\Standalone.py", line 528, in detectUsedDLLs
    _detectUsedDLLs(
    ~~~~~~~~~~~~~~~^
        standalone_entry_point=standalone_entry_point, source_dir=source_dir
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "C:\Users\xxxx\Documents\nuitka-zig\.venv\Lib\site-packages\nuitka\freezer\Standalone.py", line 412, in _detectUsedDLLs
    used_dll_paths = _detectBinaryDLLs(
        is_main_executable=standalone_entry_point.kind == "executable",
    ...<6 lines>...
        update_cache=not shallNotStoreDependsExeCachedResults(),
    )
  File "C:\Users\xxxx\Documents\nuitka-zig\.venv\Lib\site-packages\nuitka\freezer\Standalone.py", line 164, in _detectBinaryDLLs
    return detectBinaryPathDLLsWin32(
        is_main_executable=is_main_executable,
    ...<6 lines>...
        update_cache=update_cache,
    )
  File "C:\Users\xxxx\Documents\nuitka-zig\.venv\Lib\site-packages\nuitka\freezer\DllDependenciesWin32.py", line 150, in detectBinaryPathDLLsWin32
    scan_dirs = _getScanDirectories(
        package_name=package_name, original_dir=original_dir, use_path=use_path
    )
  File "C:\Users\xxxx\Documents\nuitka-zig\.venv\Lib\site-packages\nuitka\freezer\DllDependenciesWin32.py", line 236, in _getScanDirectories
    scan_dirs.extend(getSubDirectoriesWithDlls(msvc_redist_path))
                     ~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^
  File "C:\Users\xxxx\Documents\nuitka-zig\.venv\Lib\site-packages\nuitka\utils\FileOperations.py", line 600, in getSubDirectoriesWithDlls
    for dll_sub_directory in _getSubDirectoriesWithDlls(path):
                             ~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^
  File "C:\Users\xxxx\Documents\nuitka-zig\.venv\Lib\site-packages\nuitka\utils\FileOperations.py", line 607, in _getSubDirectoriesWithDlls
    for sub_directory in getSubDirectories(path=path, ignore_dirs=("__pycache__",)):
                         ~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\xxxx\Documents\nuitka-zig\.venv\Lib\site-packages\nuitka\utils\FileOperations.py", line 478, in getSubDirectories
    for root, dirnames, _filenames in os.walk(path):
                                      ~~~~~~~^^^^^^
  File "C:\Users\xxxx\AppData\Local\Python\pythoncore-3.13-64\Lib\os.py", line 351, in walk
    stack = [fspath(top)]
             ~~~~~~^^^^^
TypeError: expected str, bytes or os.PathLike object, not bool
Nuitka-Reports: Compilation crash report written to file 'nuitka-crash-report.xml'. Please include it in your bug report.
```

The source file is barebone:
```
# nuitka-project: --onefile
# nuitka-project: --zig

print('Hello, World!')
```

[nuitka-crash-report.xml](https://github.com/user-attachments/files/24168777/nuitka-crash-report.xml)

EDIT: Forgot to mention, if I remove the `--onefile` argument it works.

**Comment by kayhayen** [comment]
The bit about onefile not doing it, makes sense, but the redist path being one ought to be valid and allowed, but apparently is not properly handled. I think I touched it recently to add some caching, which was lacking there, and it so happens that once the caching kicks in the negative case, "False" can be returned instead of "None". I don't have a MSVC free system, so thanks for sharing, that probably means I don't have to uninstall MSVC which would be rather painful for me.

**Comment by kayhayen** [comment]
So, I fixed that, however, factory is still seeing unstable scalability changes, and/or repair attempts, but I am hoping to get it stable these days.

**Comment by AxelVoitier** [comment]
I can confirmed current factory fixed the onefile issue on the simple hello world test 👍 .

I will keep testing on the bigger app now.

**Comment by AxelVoitier** [comment]
As far as I can test/try, everything seems fine with the zig compilation on the bigger app now (pyside6 based fyi).

Let me know if there are more things you would like me to test on that non-msvc machine.

**Comment by kayhayen** [comment]
I have added zig on Windows for my internal CI with 3.13 and 3.14 and will see how well that fares, but I believe it cannot be that bad, obviously all C compilers have bugs, but rarely do they affect Nuitka anyway, lets hope this is not an exception. The code is still on factory only, but I am hoping to make a pre-release and release of 4.0 soon.

**Comment by fleimgruber** [comment]
I want to contribute another test data point of a project that uses 20+ dependencies including Pandas and it all works flawlessly with current `factory` branch. Also happy to do more testing.

**Comment by jimkring** [comment]
@fleimgruber that's great news on your success. Thanks for sharing and adding to the momentum!

**Comment by kayhayen** [comment]
I have added a build for basic tests in my private CI and it works on the 3 major OSes. I am a bit worried about issues with --debug` for maintenance, but it is what it is.

**Comment by kayhayen** [comment]
This is part of the stable release 4.0 that I just made.
